### PR TITLE
migrating functions from Jupyter notebook to IDE

### PR DESCRIPTION
**Pull request by WellAndGood**
Moving methods, definitions from functioning Jupyter notebook. Now on Visual Studio Code.
Incomplete.

### CODE REVIEW

File: AP_article_builder.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import pprint
+import csv
+import datetime
+import nltk
+import requests
+import csv
+import json
+import spacy
+from spacy import displacy
+import pprint
```
Comment (rfazeli) [inline AP_article_builder.py]: Please remove all unused imports. Do the same for all files.

File: AP_article_builder.py
Comment (rfazeli) [inline AP_article_builder.py:1]: Python script names should be all lowercase.

File: AP_article_builder.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+# Works for Associated Press Articles
+def AP_article_dict_builder(url):
```
Comment (rfazeli) [inline AP_article_builder.py]: Python function names should be all lowercase. This applies to all function names in your scripts.

Comment (rfazeli) [reply AP_article_builder.py]: It'd be great if you could add docstrings and typing for your functions as well. See the example below:
```python
from typing import List

def str_to_list(string: str) -> List[str]:
        """
        Static method that converts a string with items 
        separated by \n and converts it to a list

        Parameters
        ----------
        string : str
            The input string

        Returns
        -------
        List[str]
            The list of ideas
        """

        items = string.strip("\n").strip().split("\n")
        items = [item.strip() for item in items]
        return items
```

File: AP_article_builder.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    # Takes URL, outputs Beautiful Soup html
+    response = requests.get(url)
+    APArticle = BeautifulSoup(response.content, "html.parser")
```
Comment (rfazeli) [inline AP_article_builder.py]: Make sure your variable names follow snake casing which is the convention for python i.e. words separated by underscores ( _ ), with all letters in lowercase.

`APArticle -> ap_article`

This comment applies to all variable names in your project.

File: AP_article_builder.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    article_information = {}
+
+    # Retrieves canonical link
+    canonical_links = APArticle.find_all("link", rel="canonical")
+
+    # Print the href attribute of each matching link
+    for link in canonical_links:
+        self_URL = link.get("href")
```
Comment (rfazeli) [inline AP_article_builder.py]: This for loop seems to be only storing the last url. Why don't you just do 
```python
self_url = canonical_links[-1].get("href")
```

File: AP_article_builder.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    article_information["author(s)"] = author_list
+
+    # Retrieves image data
+    image_data = []
+    image_url = JSONscript["image"]
+    image_caption_div = APArticle.find_all("div", attrs={"data-key": "embed-caption"})
+    image_caption = image_caption_div[0].text
+    try:
+        image_attribution = image_caption[-50:].replace(")", "").split("(")[1]
+    except:
```
Comment (rfazeli) [inline AP_article_builder.py]: It'd be good to catch the specific exception. I'm assuming in this case it'd be a `IndexError` so you can change this line to
```python
except IndexError:
```

File: db_interaction.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+# Creates a unique hash of a string
+def hash_string(string):
+    hash_object = hashlib.md5()
+    hash_object.update(string.encode())
+    return hash_object.hexdigest()
+
+
+# Sentence list
+url = "https://apnews.com/article/george-santos-federal-charges-updates-33667a0900271e5002459ab748d8fdc8?utm_source=homepage&utm_medium=TopNews&utm_campaign=position_01"
```
Comment (rfazeli) [inline db_interaction.py:30]: Is this a constant? If yes you can have a constants.py file and add all constants there. And constant are all uppercase. So it'd be:
```python
URL = "https://apnews.com/article/george-santos-federal-charges-updates-33667a0900271e5002459ab748d8fdc8?utm_source=homepage&utm_medium=TopNews&utm_campaign=position_01"
```

File: main.py [hunk 7]
```diff
@@ -1,24 +1,34 @@
...
-url_pattern = re.compile(
-    r'^(https?|ftp)://'  # Protocol (http, https, ftp)
-    r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # Domain
-    r'(?:/?|[/?]\S+)$', re.IGNORECASE  # Ignore case
-)
+# Initialize the Doc object
+doc = nlp(article_txt)
+
+sentences = sentence_generator(doc)
+verbs = verb_matcher(doc)
```
Comment (rfazeli) [inline main.py]: You should move all of this under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+# https://apnews.com/article/george-santos-federal-charges-updates-33667a0900271e5002459ab748d8fdc8?utm_source=homepage&utm_medium=TopNews&utm_campaign=position_01
+
+# Base URL, and use
+url = "https://apnews.com/article/george-santos-federal-charges-updates-33667a0900271e5002459ab748d8fdc8?utm_source=homepage&utm_medium=TopNews&utm_campaign=position_01"
+article_dict = AP_article_dict_builder(url)
+article_txt = AP_article_full_txt(url)
+
+# Initialize the Doc object
+doc = nlp(article_txt)
```
Comment (rfazeli) [inline spacy_methods.py]: What is this part of the code? If it's an example for using the functionality in this script you should move them under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                            word_index,
+                        ]
+                    )
+
+            # Adds the sentence_end indexing to the total word count
+        word_count += len(doc)
+    return specific_entities
+
+
+entities = get_specific_entities(sentences)
```
Comment (rfazeli) [inline spacy_methods.py:73]: Same comment about moving example code under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+# raw_entity_list = list(entities)
```
Comment (rfazeli) [inline spacy_methods.py]: You can remove this if it's not needed

File: spacy_methods.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+        entity_sentence_check = "{}".format(item[0])
+        if entity_sentence_check in count_dict:
+            count_dict[entity_sentence_check] += 1
+        else:
+            count_dict[entity_sentence_check] = 1
+    return count_dict
+
+
+duplicate_items = entity_counter(entities)
```
Comment (rfazeli) [inline spacy_methods.py:92]: Same comment about moving it under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+verbs = verb_matcher(doc)
+# print(verbs)
+
+# List of Verbs
+verbs = verb_matcher(doc)
+
+# List of Article's Sentences
+
+sentences = sentence_generator(doc)
```
Comment (rfazeli) [inline spacy_methods.py]: Same comment about moving it under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+                        )
+                    )
+        except:
+            continue
+
+    return specific_verbs
+
+
+the_verbs = verb_in_sentence(verbs, sentences)
+ppv = pprint.PrettyPrinter(indent=4)
```
Comment (rfazeli) [inline spacy_methods.py]: Same comment about moving it under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+        entity_sentence_check = "{}".format(entity_item[0])
+
+        # entity_item[0] - Entity's name as a string
+        # entity_item[5] - the sentence number index
+
+        append_to_array(entity_item[0], entity_item[5], ent_index_dict)
+    return ent_index_dict
+
+
+ent_sentence_index = entity_indexer(raw_entity_list)
```
Comment (rfazeli) [inline spacy_methods.py]: Same comment about moving it under `if __name__ == "__main__":`

File: spacy_methods.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+# Algorithm to assign new key values to 1, increment if already existing
+def append_to_array(key, value, dict_to_check):
+    if key in dict_to_check:
+        dict_to_check[key].append(value)
+    else:
+        dict_to_check[key] = [value]
+
+
+raw_entity_list = list(entities)
```
Comment (rfazeli) [inline spacy_methods.py:102]: Same comment about moving it under `if __name__ == "__main__":`

**Review by rfazeli — CHANGES_REQUESTED** [review]
Thanks very much for this PR Daniel. I think this is a great start. I've added some comments inline. 

One general comment: I think you can move the scripts that provide certain functionality under an `src/` directory and keep the run scripts (where you import those functions and use them to perform a certain task) in the root dir. Just make sure to update those imports to include src. as well

File: AP_article_builder.py (hunk 1 above)
Comment (WellAndGood) [reply AP_article_builder.py]: Resolved in my local repo.

File: AP_article_builder.py (hunk 2 above)
Comment (WellAndGood) [reply AP_article_builder.py]: Functions changed to lowercase,. Typing added; not documented. Pushing

File: AP_article_builder.py (hunk 4 above)
Comment (WellAndGood) [reply AP_article_builder.py]: Fixed.

File: AP_article_builder.py (hunk 3 above)
Comment (WellAndGood) [reply AP_article_builder.py]: Fixed.

File: spacy_methods.py (hunk 12 above)
Comment (WellAndGood) [reply spacy_methods.py]: Moved. Resolved.

File: spacy_methods.py (hunk 11 above)
Comment (WellAndGood) [reply spacy_methods.py:92]: Moved. Resolved.

File: spacy_methods.py (hunk 14 above)
Comment (WellAndGood) [reply spacy_methods.py]: Moved. Resolved.

File: spacy_methods.py (hunk 13 above)
Comment (WellAndGood) [reply spacy_methods.py]: Moved. Resolved.

File: AP_article_builder.py
Comment (WellAndGood) [reply AP_article_builder.py:1]: Moved. Resolved.

File: AP_article_builder.py (hunk 5 above)
Comment (WellAndGood) [reply AP_article_builder.py]: Fixed. Resolved

File: main.py (hunk 7 above)
Comment (WellAndGood) [reply main.py]: Sure. Done.

File: spacy_methods.py (hunk 10 above)
Comment (WellAndGood) [reply spacy_methods.py]: Removed.

File: spacy_methods.py (hunk 9 above)
Comment (WellAndGood) [reply spacy_methods.py:73]: Moved.

File: spacy_methods.py (hunk 8 above)
Comment (WellAndGood) [reply spacy_methods.py]: The url was to pass through an example. I've created the constants.py file, and added two URL constants.

File: db_interaction.py (hunk 6 above)
Comment (WellAndGood) [reply db_interaction.py:30]: This url was to pass through an example. I've created the constants.py file, and added two URL constants. One is the main AP News website. The second is the URL above.

**Comment by WellAndGood** [comment]
Changes implemented. Core features of NLP analysis completed. Pull request completed.

**Pull request merged**